bext: update private repo sync banners

taras-yemets opened the pull request:
Closes [issues/31920](https://github.com/sourcegraph/sourcegraph/issues/31920).

See also PR https://github.com/sourcegraph/sourcegraph/pull/31690 discussion.

Show different instructions on the options page for repo not found error depending on Sourcegraph URL and whether the user is admin.

| Sourcegraph instance  | User is admin | Screenshot | Notes |
| ------------------------------ | ------------------ | ---------------- | -------- |
| Cloud                           |                        |  ![Screenshot from 2022-03-24 15-38-02](https://user-images.githubusercontent.com/25318659/160400906-9be3f796-2332-448a-acdc-16accb7488f3.png)   |  Sourceraph button text is 'Configure on Sourcegraph'. It navigates to user settings manage repositories page with current repo name as filter query.         |
| Self-hosted / managed |  ✅                 |  ![Screenshot from 2022-03-24 14-23-46](https://user-images.githubusercontent.com/25318659/160401284-37a6f3f5-66e5-4590-af22-622f706e8a9e.png)  | - Link in the extension popup navigates user to [add repos docs page](https://docs.sourcegraph.com/admin/repo/add)<br> - Sourceraph button text is 'Repository not found'. It still [links to the repository](https://github.com/sourcegraph/sourcegraph/blob/main/client/browser/src/shared/code-hosts/shared/ViewOnSourcegraphButton.tsx#L134), but will show 'Not Found' page. This information is just for context, the button is not the matter of change of this PR. |
| Self-hosted / managed |  ❌                 | ![Screenshot from 2022-03-24 15-37-27](https://user-images.githubusercontent.com/25318659/160401219-590f00d7-7e36-4900-93ba-e1b0e4979153.png)  | ⬆️ See notes on the Sourcegraph button for the on-prem / managed instance admin user. | 
 
UPD: updated copies according to [this comment](https://github.com/sourcegraph/sourcegraph/pull/31922#issuecomment-1081577004)
![repo sync](https://user-images.githubusercontent.com/25318659/160581974-08ab9cd2-16e7-4681-bd4c-626954fb36a8.png)


## Test plan
 - `sg run bext`
 - Sourcegraph cloud instance
   - set Sourcegraph URL to Sourcegrpah Cloud
   - log in to your cloud account
   - go to the private repository page which is not synced in user settings 
   - check the private repo message on the options page (see screenshots above)
 - Sourcegraph on-prem / managed instance
   - set Sourcegrpah URL to on-prem / managed instance
   - log in as non-admin user
   - check not-synced repo message on the options page (see screenshots above)
   - log in as admin user
   - check not-synced repo message for the admin user (see screenshots above)


## App preview:
- [Link](https://sg-web-taras-yemets-bext-update-private.onrender.com)

tjkandala commented on client/browser/src/browser-extension/scripts/optionsPage.main.tsx:
```diff
@@ -140,6 +141,25 @@ function useTelemetryService(sourcegraphUrl: string | undefined): TelemetryServi
+        variables: {},
+    }).pipe(
+        map(({ data }) => data?.currentUser?.settingsURL),
+        filter(Boolean),
+        map(url => new URL(`${url as string}/repositories/manage`, sourcegraphURL).href)
```
```suggestion
        filter(isDefined),
        map(url => new URL(`${url}/repositories/manage`, sourcegraphURL).href)
```

**Suggestion**: We should be able to use the [`isDefined`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/common/src/types/utils.ts?L4%3A20=&utm_source=VSCode-2.2.0) type guard here so we can remove the type assertion. There's little to no risk of `settingsURL` type changing, but it would be ideal to use type guards to safeguard against that consistently (sad that TS can't infer from `filter(Boolean)` correctly, though)

taras-yemets replied:
I like this suggestion! Thanks @tjkandala!

tjkandala commented on client/browser/src/browser-extension/scripts/optionsPage.main.tsx:
```diff
@@ -232,9 +261,7 @@ const Options: React.FunctionComponent = () => {
-                    showPrivateRepositoryAlert={
-                        currentTabStatus?.status.hasPrivateCloudError && isDefaultSourcegraphUrl(sourcegraphUrl)
-                    }
+                    manageRepositoriesURL={manageRepositoriesURL}
```
Nice refactoring

taras-yemets commented on client/browser/src/shared/code-hosts/shared/ViewOnSourcegraphButton.tsx:
```diff
@@ -60,16 +62,13 @@ export const ViewOnSourcegraphButton: React.FunctionComponent<ViewOnSourcegraphB
-
     useEffect(() => {
-        onPrivateCloudError(isPrivateCloudError)
+        onPrivateCloudError(repoExistsOrError === false)
```
Notify about repo sync error if repo does not exist.
TODO: rename `privateCloudError` and its usages to `repoSyncError` or something like that.

taras-yemets commented on client/browser/src/browser-extension/scripts/backgroundPage.main.ts:
```diff
@@ -53,10 +53,11 @@ initSentry('background')
+ * it has experienced repository does not exist on the instance error
+ * by sending `notifyRepoSyncError` message.
  */
 const tabPrivateCloudErrorCache = (() => {
```
Convert `tabPrivateCloudErrorCache` to `tabRepoSyncErrorCache`. 

Now it keeps track of tab ids of:
- private repositories not synced with Sourcegraph Cloud when browser extension has _sourcegraph.com_ Sourcegrpah URL (existing functionality)
- repositories not added to other than Sourcegraph Cloud instance, for example, _k8s.sgdev.org_ (new functionality) 

TODO: rename `privateCloudError` and its usages to `repoSyncError` or something like that.

erzhtor commented on client/browser/src/browser-extension/scripts/optionsPage.main.tsx:
```diff
@@ -153,15 +174,22 @@ const Options: React.FunctionComponent = () => {
+                ),
+            []
+        )
+    )
```
Nice improvement! 👍

erzhtor commented on client/browser/src/browser-extension/scripts/optionsPage.main.tsx:
```diff
@@ -221,6 +249,7 @@ const Options: React.FunctionComponent = () => {
     return (
         <ThemeWrapper>
             <WildcardThemeProvider isBranded={true}>
+                {sourcegraphUrl}/ error: {currentTabStatus?.status?.hasRepoSyncError.toString()}
```
Is this debug code?

erzhtor commented on client/browser/src/browser-extension/options-menu/OptionsPage.tsx:
```diff
@@ -187,34 +195,56 @@ const PermissionAlert: React.FunctionComponent<PermissionAlertProps> = ({
-        </ol>
-    </section>
-)
+const RepoSyncErrorAlert: React.FunctionComponent<Pick<OptionsPageProps, 'sourcegraphUrl' | 'currentUser'>> = ({
```
Considering currentUser checking moved outside, we can make types more explicit:
```jsx
const RepoSyncErrorAlert: React.FunctionComponent<NonNullable<Pick<OptionsPageProps, 'sourcegraphUrl' | 'currentUser'>>> = ({
```

erzhtor commented on client/browser/src/shared/code-hosts/shared/ViewOnSourcegraphButton.tsx:
```diff
@@ -60,16 +62,13 @@ export const ViewOnSourcegraphButton: React.FunctionComponent<ViewOnSourcegraphB
-
     useEffect(() => {
-        onPrivateCloudError(isPrivateCloudError)
+        onRepoSyncError(repoExistsOrError === false)
```
Was this an intentional change, and won't it somehow affect previous logic?

taras-yemets replied:
Yes, this change is intentional. It doesn't seem to affect buttons rendering logic (will double-check it after finalizing the PR).
Motivation for this change is described in [this comment](https://github.com/sourcegraph/sourcegraph/pull/31922#discussion_r834057215).

erzhtor commented on client/browser/src/shared/code-hosts/shared/ViewOnSourcegraphButton.tsx:
```diff
@@ -35,21 +36,22 @@ interface ViewOnSourcegraphButtonProps
     onSignInClose?: () => void
 
-    onPrivateCloudError: (hasError: boolean) => void
+    onRepoSyncError: (hasError: boolean) => void
```
WDYT to add jsDocs here on what is this and what it does?

erzhtor commented on client/browser/src/shared/code-hosts/shared/ViewOnSourcegraphButton.tsx:
```diff
@@ -116,18 +115,12 @@ export const ViewOnSourcegraphButton: React.FunctionComponent<ViewOnSourcegraphB
+        }
+
         return (
             <SourcegraphIconButton
```
Just a random suggestion, maybe we could change the title for this case from "Repository not found" to "Repository not synced," WDYT? But worth checking with Rob/Sara anyways.

ryankscott replied:
These mean subtlety different things, how I read it: 
- Not found - meaning that Sourcegraph has no knowledge of this repository (like a 404)
- Not synced - meaning that Sourcegraph is aware of the repository but has imported / synced it to Sourcegraph.

I believe it's the first one and consequently we should say "Repository not found"

erzhtor commented on client/browser/src/shared/code-hosts/shared/codeHost.tsx:
```diff
@@ -930,7 +935,7 @@ export async function handleCodeHost({
         hoverAlerts,
         mutations,
-        privateCloudErrors,
+        privateCloudErrors: repoSyncErrors,
```
Should we drill rename "privateCloudErros" here too?

erzhtor reviewed (approved):
Great job, and thank you for making this change! 💪  Looking forward to seeing this shipped to production! 🚀

taras-yemets commented:
@erzhtor, thank you for your review: lots of useful comments! 
Will address them after I prove that everything works as intended and that this implementation actually _solves_ problems rather than introduces new ones. These changes seemed trivial at first, but now I see that there could be unintended side-effects here and there.
I also faced some browser extension weird behavior while testing this (for example, [this one](https://www.loom.com/share/8f82e8438c8942089c0e0022603fb8f8?sharedAppSource=personal_library)). So I'm going to get all of these sorted and then proceed with polishing the PR.
Thanks again!

ryankscott commented:
Suggested some minor wording changes. I've tried my best to keep it to our [content guidelines](https://handbook.sourcegraph.com/company-info-and-process/communication/content_guidelines/). The links look correct, I suggest we put them where the bold text is below.

![image](https://user-images.githubusercontent.com/2660760/160564262-4de44db2-5f4b-4449-a56d-d04a7c809fab.png)
- Update title to say "Repository not found"
- Update description to say "**Add your repository to Sourcegraph** to use this extension." 

![image](https://user-images.githubusercontent.com/2660760/160564878-942951b6-3a2b-46bd-8a60-c1a1c72a2628.png)
- Update description to say "**Add your repository to Sourcegraph** to use this extension for private repositories."

![](https://user-images.githubusercontent.com/25318659/160401219-590f00d7-7e36-4900-93ba-e1b0e4979153.png)
- Update title to say "Repository not synced" (Or not found depending on which it is, see above comment)
- Update description to say: "Contact your admin to sync this repository with Sourcegraph"



As a future point, we should help users identify their admin as it's likely in a big company you might not know who this is.

taras-yemets commented:
> Suggested some minor wording changes. 

Addressed these comments and updated the PR description.

> * Update title to say "Repository not synced" (Or not found depending on which it is, see above comment)
> * Update description to say: "Contact your admin to sync this repository with Sourcegraph"

Used "Repository not found" title and "Contact your admin to add this repository to Sourcegraph." text for on-prem non-admin users to be consistent with the similar case for admin users.

> As a future point, we should help users identify their admin as it's likely in a big company you might not know who this is.

Good point!

ryankscott commented:
LGTM

erzhtor commented on client/browser/src/browser-extension/scripts/backgroundPage.main.ts:
```diff
@@ -53,33 +53,53 @@ initSentry('background')
-            if (!hasPrivateCloudError) {
-                // An absent value is equivalent to being false; so we can delete it.
+        setTabHasRepoSyncError(tabId: number, sourcegraphURL: string, hasRepoSyncError: boolean): void {
+            if (sourcegraphURL) {
```
In this case, can we make this parameter type optional to make it explicit?

taras-yemets replied:
Agree!

erzhtor commented on client/browser/src/browser-extension/scripts/backgroundPage.main.ts:
```diff
@@ -53,33 +53,53 @@ initSentry('background')
-            if (!hasPrivateCloudError) {
-                // An absent value is equivalent to being false; so we can delete it.
+        setTabHasRepoSyncError(tabId: number, sourcegraphURL: string, hasRepoSyncError: boolean): void {
+            if (sourcegraphURL) {
+                const record = cache.get(tabId)
+
+                if (record) {
+                    if (hasRepoSyncError) {
+                        record.set(sourcegraphURL, true)
+                    } else {
+                        record.delete(sourcegraphURL)
+                    }
+
+                    if (record.size === 0) {
+                        cache.delete(tabId)
+                    }
+                } else if (hasRepoSyncError) {
+                    cache.set(tabId, new Map().set(sourcegraphURL, true))
+                }
+            } else {
                 cache.delete(tabId)
             }
```
Just minor rearranging to decrease if/else nesting:
```suggestion
            if (sourcegraphURL) {
                let record = cache.get(tabId)

                if (!record) {
                    record = new Map()
                    cache.set(tabId, record)
                }

                if (hasRepoSyncError) {
                    record.set(sourcegraphURL, true)
                } else if (record.size === 0) {
                    cache.delete(tabId)
                } else {
                    record.delete(sourcegraphURL)
                }
            } else {
                cache.delete(tabId)
            }
```

taras-yemets replied:
Nice refactoring!

erzhtor commented on client/browser/src/browser-extension/scripts/backgroundPage.main.ts:
```diff
@@ -53,33 +53,53 @@ initSentry('background')
+                    cache.set(tabId, new Map().set(sourcegraphURL, true))
+                }
+            } else {
                 cache.delete(tabId)
```
*This is not changed in your PR, but something that I noticed.*

Based on this logic and when/where it is called, I think, we never cleanup closed tabs which in turn might blow memory usage by Browser Extension background script. We need to add corresponding cleanup logic, I believe.
```js
    browser.tabs.onRemoved.addListener(tabId => {
        tabRepoSyncErrorCache.setTabHasRepoSyncError(tabId, '', false) // this removes cache item for a tabId
    })
```

taras-yemets replied:
Nice catch! Thanks for fixing it!

erzhtor commented on client/browser/src/browser-extension/scripts/backgroundPage.main.ts:
```diff
@@ -53,33 +53,53 @@ initSentry('background')
-        setTabHasPrivateCloudError(tabId: number, hasPrivateCloudError: boolean): void {
-            if (!hasPrivateCloudError) {
-                // An absent value is equivalent to being false; so we can delete it.
+        setTabHasRepoSyncError(tabId: number, sourcegraphURL: string, hasRepoSyncError: boolean): void {
```
To clarify, why do we need this `sourcegraphURL`? Is this is due to Browser Extension supporting in page sourcegraphURL change without needing to reload the tab?

taras-yemets replied:
Yes, correct, it's because a user can change the URL without page reloading.

erzhtor commented on client/browser/src/browser-extension/scripts/backgroundPage.main.ts:
```diff
@@ -175,7 +195,7 @@ async function main(): Promise<void> {
         if (changeInfo.status === 'loading') {
             // A new URL is loading in the tab, so clear the cached private cloud error flag.
-            tabPrivateCloudErrorCache.setTabHasPrivateCloudError(tabId, false)
+            tabRepoSyncErrorCache.setTabHasRepoSyncError(tabId, '', false)
```
if we make sourcegraphURL optional and 3rd parameter, we can avoid passing it at all here:
```js
tabRepoSyncErrorCache.setTabHasRepoSyncError(tabId, false)
```

erzhtor commented on client/browser/src/shared/backend/requestGraphQl.ts:
```diff
@@ -16,7 +16,6 @@ interface RequestGraphQLOptions<V> {
     request: string
     variables: V
     mightContainPrivateInfo: boolean
-    privateCloudErrors?: Observable<boolean>
```
To clarify, this was never used really? Is this why we can safely remove it?

taras-yemets replied:
Yes, I've been looking for `privateCloudErrors` usages but couldn't find any.

erzhtor commented on client/browser/src/shared/code-hosts/shared/codeHost.tsx:
```diff
@@ -1001,46 +1005,58 @@ export async function handleCodeHost({
-                        // The bound function is constant
-                        onSignInClose={nextSignInClose}
-                        onConfigureSourcegraphClick={isInPage ? undefined : onConfigureSourcegraphClick}
-                        onPrivateCloudError={onPrivateCloudError}
```
👍 

P.S: If I correctly understood, this was a duplicate logic already being handled before this place, right?

taras-yemets replied:
We moved the repo sync error notification logic [outside the component](https://github.com/sourcegraph/sourcegraph/blob/3f2d1ebfbcafca0c32ae86902738805a9bd77079/client/browser/src/shared/code-hosts/shared/codeHost.tsx#L1009-L1019). Now we notify about this kind of error regardless whether we render `<ViewOnSourcegraphButton />` (`codeHost.getViewContextOnSourcegraphMount` is defined) or not.
We made this change to track repo sync errors where `codeHost.getViewContextOnSourcegraphMount` is not defined we don't render `<ViewOnSourcegraphButton />` (for example, on GitLab).

erzhtor reviewed (approved):
Tremendous work done! 👍 Thank you for tackling and addressing comments!

erzhtor commented:
@taras-yemets, I've pushed a commit w/ minor changes/fixes, primarily addressing my comments above. Please, feel free to discard/update as you think is better.
